examples/xds-management-server/pkg/resources/resources.go

```diff
@@ -0,0 +1,199 @@
+		[]types.Resource{}, // endpoints
+		clusterResources,
+		[]types.Resource{}, // routes
+		[]types.Resource{listener}, // listeners
+		[]types.Resource{}, // runtimes
+		[]types.Resource{}, // secrets
+	), nil
+}
+
+func makeCluster(clusterName string, cluster Cluster) (*envoycluster.Cluster, error) {
+	loadAssignment, err := makeEndpoint(clusterName, cluster.Endpoints)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cluster resource: %w", err)
+	}
+
+	return &envoycluster.Cluster{
+		Name: clusterName,
+		ClusterDiscoveryType: &envoycluster.Cluster_Type{
+			Type: envoycluster.Cluster_STATIC,
+		},
+		LoadAssignment: loadAssignment,
+	}, nil
+}
+
+func parseMetadata(input map[string]interface{}) (map[string]*structpb.Struct, error) {
+	output := make(map[string]*structpb.Struct)
+	for key, value := range input {
+		metadataBytes, err := json.Marshal(value)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse metadata: %w", err)
+		}
+
+		protoValue := structpb.Struct{}
+		if err := jsonpb.Unmarshal(bytes.NewReader(metadataBytes), &protoValue); err != nil {
+			return nil, err
+		}
+		output[key] = &protoValue
+	}
+	return output, nil
+}
+
+func makeEndpoint(
+	clusterName string,
+	endpoints []Endpoint,
+) (*envoyendpoint.ClusterLoadAssignment, error) {
+	var endpointConfigs []*envoyendpoint.LocalityLbEndpoints
+	for _, ep := range endpoints {
+		metadata, err := parseMetadata(ep.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create Endpoint resource: %w", err)
+		}
+		endpointConfigs = append(endpointConfigs, &envoyendpoint.LocalityLbEndpoints{
+			LbEndpoints: []*envoyendpoint.LbEndpoint{{
+				Metadata: &envoycore.Metadata{
+					FilterMetadata: metadata,
+				},
+				HostIdentifier: &envoyendpoint.LbEndpoint_Endpoint{
+					Endpoint: &envoyendpoint.Endpoint{
+						Address: &envoycore.Address{
+							Address: &envoycore.Address_SocketAddress{
+								SocketAddress: &envoycore.SocketAddress{
+									Protocol: envoycore.SocketAddress_UDP,
+									Address:  ep.IP,
+									PortSpecifier: &envoycore.SocketAddress_PortValue{
+										PortValue: uint32(ep.Port),
+									},
+								},
+							},
+						},
+					},
+				},
+			}},
+		})
+	}
+
+	return &envoyendpoint.ClusterLoadAssignment{
+		ClusterName: clusterName,
+		Endpoints:   endpointConfigs,
+	}, nil
+}
+
+func makeListener(
+	filterChainResource []FilterConfig,
+) (*envoylistener.Listener, error) {
+	filterChain, err := makeFilterChain(filterChainResource)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create listener resource: %w", err)
+	}
+
+	return &envoylistener.Listener{
+		FilterChains: []*envoylistener.FilterChain{ filterChain },
+	}, nil
+}
+
+func makeFilterChain(
+	filterChainResource []FilterConfig,
+) (*envoylistener.FilterChain, error) {
+	var filters []*envoylistener.Filter
+
+	for _, resource := range filterChainResource {
+		switch resource.Name {
+		case "quilkin.extensions.filters.debug.v1alpha1.Debug":
```

Comment:
🤔 I wonder if there is a way to make this generic without having to be explicit about type.

Just figuring that If this has a static list of Filters and accompanying models here, then each time a user / we release a new build of Quilkin with a new Filter, we have to update the xDS code and deployment -- if we could be generic here, it might be easier in the long term.

But I can't think of a good way to do that 😦 as per #346 maybe we can generate some kind of go map/yaml/json with accompanying code or something that makes this piece of code do a lookup, rather than have to implement a giant switch. (Thinking out loud at this point).

Reply:
I did some digging and reworked this now PTAL! now we capture filter configs as a generic `interface` object and marshal it directly into an xds filter so we don't need to know which particular filter it is (and not switching/lookup to do either). Only needed to [change the config to match the xds proto](https://github.com/googleforgames/quilkin/blob/0d9a3111d1975c8c1a2ccafce00de83ceb1c73b5/examples/xds-management-server/config.json#L17-L23) in order to make it directly serializable but its the same format istio uses as well to describe envoy filter configs so its not too awkward at least https://istio.io/latest/docs/reference/config/networking/envoy-filter/